fields/ActivityNavigationField.php

```diff
@@ -64,8 +67,27 @@ protected function renderStatic($entry)
 
         $output = '';
         if ($course && $module && !empty($module->getActivities())) {
-            // save the activity progress if not already exists for this user and activity
-            $this->learnerManager->saveActivityProgress($course, $module, $activity);
+            $learner = $this->learnerManager->getLearner();
+            if ($learner
+                && $this->courseManager->setModuleCanBeOpenedByLearner(
+                    $learner,
+                    $course,
+                    $module
+                ) // module should be accessible
+                && ($referenceActivity = $module->getActivity($activity->getTag()))
+                 // activity should be in module
+                && !is_null($this->courseManager->setActivityCanBeOpenedByLearner(
+                    $learner,
+                    $course,
+                    $module,
+                    $referenceActivity
+                )) // set Activity Can be opened
+                && !is_null($activity->canBeOpenedBy($learner, $referenceActivity->canBeOpenedBy($learner)))
+                && $activity->isAccessibleBy($learner, $course, $module)
```

Comment:
dans le cas où si le module ou l'activité ne sont pas accessibles, Ok pour ne pas enregistrer mais je pense qu'il ne faut aussi n'afficher aucun bouton.

au final, il suffit de faire `$course && $module && $module->isAccessibleBy($learner, $course) && $activity->isAccessibleBy($learner, $course)` car getContextualModule suffit à vérifier que $activity fait partie de $module.
Je n'ai pas compris pourquoi tu faisais $referenceActivity = $module->getActivity($activity->getTag()).

Reply:
>$referenceActivity = $module->getActivity($activity->getTag()).

en fait, c'est un contournement temporaire du fait que je ne comprends pas trop pourquoi on instancie une nouvelle activity ici sans la demander au courseManager dont c'est le rôle.

Reply:
pourtant le commentaire me semblait déjà un bon début de réponse :

> // the activity is not loaded from the manager because we don't want to requests the fields (it's an exception)

j'ai fais tout un travail d'optimisation des chargements car connaissant yeswiki et ses tendances à ramer lorsque les instances ont bcp de données (cf les sites de mooc avec des milliers d'utilisateurs), il y a certains petits hacks qui sont voulus. Avec le Mooc TI, on voit qu'on a plus les mêmes pb de perfs d'avant, grâce j'imagine à twig et son système de cache.

dans ce cas précis, vu que la page est déjà chargée, elle est déjà en cache donc le gain est peut-être négligeable mais j'avais préféré assurer. Mais le hack par exemple pour charger la liste des activités sans charger chacunes d'elles me semble primordial.

pareil, y'a des chargements systématiques que tu rajoutes avec la scénarisation qu'il faudrait voir s'ils sont vraiment nécessaires.

Reply:
corrigé ->https://github.com/YesWiki/yeswiki-extension-lms/commit/6309eae425f2ac5497562400d261b1766a3a18b7